sensor/common/detector/baseline/baseline.go

```diff
@@ -88,3 +102,170 @@ func (w *baselineEvaluator) IsOutsideLockedBaseline(pi *storage.ProcessIndicator
 	// If there is no baseline, then we are counting it as if it's within the baseline
 	return baseline != nil && !baseline.Contains(processbaseline.BaselineItemFromProcess(pi))
 }
+
+// optimizedBaselineEvaluator implements memory-optimized baseline evaluation using process set deduplication
+type optimizedBaselineEvaluator struct {
+	// deployment -> container name -> content hash (direct access)
+	deploymentBaselines map[string]map[string]string
+	// content hash -> reference count and StringSet for deduplication
+	processSets map[string]*processSetEntry
+	// lock for thread safety
+	lock sync.RWMutex
+}
+
+type processSetEntry struct {
+	refCount  int
+	processes set.StringSet
+}
+
+// newOptimizedBaselineEvaluator creates the optimized baseline evaluator implementation
+func newOptimizedBaselineEvaluator() Evaluator {
+	return &optimizedBaselineEvaluator{
+		deploymentBaselines: make(map[string]map[string]string),
+		processSets:         make(map[string]*processSetEntry),
+	}
+}
+
+// removeReference decrements reference count and cleans up if necessary
+func (oe *optimizedBaselineEvaluator) removeReference(contentHash string) {
+	entry := oe.processSets[contentHash]
+	if entry == nil {
+		return // Entry doesn't exist
+	}
```

Comment:
Shouldn't we maybe use the safer variant here?
What if the map has an entry that is a nil (is that possible)?


```suggestion
	entry, found := oe.processSets[contentHash]
	if !found || entry == nil {
		return // Entry doesn't exist
	}
```

Reply:
Good point, cleaned this function up.

Reply:
This was addressed in `removeReference` (thanks!), however I see similar pattern in `IsOutsideLockedBaseline`. Could we check if the behavior there is as expected?